### PR TITLE
Added initial taint view model implementations

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Relates to #1985

### CODE REVIEW

File: src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Editor;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+using SonarLint.VisualStudio.IssueVisualization.Security.Taint;
+using SonarLint.VisualStudio.IssueVisualization.Security.Taint.TaintList.ViewModels;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.Taint.TaintList
+{
+    [TestClass]
+    public class TaintIssuesControlViewModelTests
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.cs:35]: Copied from the equivalent hotspots tests. Only the types and variable names have changed.

Same for the other test class.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+using System.IO;
+using System.Runtime.CompilerServices;
+using SonarLint.VisualStudio.Core.Analysis;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Taint.TaintList.ViewModels
+{
+    internal interface ITaintIssueViewModel : INotifyPropertyChanged, IDisposable
+    {
+        IAnalysisIssueVisualization TaintIssueViz { get; }
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs:32]: The volume of code was getting too large to manage conveniently for review purposes so I changed my approach. I'm adding the necessary classes for a minimal working tool window that will display something, then add functionality incrementally (additional columns, selection, filtering etc).

Breaking up the work should make it easier to review and to identify opportunities for refactoring. We'll also have a working end to end sooner.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        int Line { get; }
+
+        int Column { get; }
+
+        string DisplayPath { get; }
+    }
+
+    internal sealed class TaintIssueViewModel : ITaintIssueViewModel
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs:41]: Copied from the hotspot view model with a few bits deleted.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                NotifyPropertyChanged(nameof(Line));
+                NotifyPropertyChanged(nameof(Column));
+            }
+            else if (e.PropertyName == nameof(IAnalysisIssueVisualization.CurrentFilePath))
+            {
+                NotifyPropertyChanged(nameof(DisplayPath));
+            }
+        }
+
+        public event PropertyChangedEventHandler PropertyChanged;
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssueViewModel.cs:81]: This event and `NotifyPropertyChanged` aren't needed yet. I`ll delete them.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+using Microsoft.VisualStudio.Shell;
+using SonarLint.VisualStudio.IssueVisualization.Editor;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Taint.TaintList.ViewModels
+{
+    internal interface ITaintIssuesControlViewModel : IDisposable
+    {
+        // WIP: add selection handling
+        // WIP: add any other necessary commands
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:38]: Again, copied from the hotspots control VM with bits taken out.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        ICommand NavigateCommand { get; }
+
+        ObservableCollection<ITaintIssueViewModel> Issues { get; }
+    }
+
+    /// <summary>
+    /// View model that surfaces the data in the TaintStore
+    /// </summary>
+    /// <remarks>Monitors the taint store for changes and updates the view model accordingly.</remarks>
+    internal sealed class TaintIssuesControlViewModel : ITaintIssuesControlViewModel
```
Comment (rita-g-sonarsource) [inline src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:49]: Looking at the code, I think the sheer amount of duplication is worthy of a generic component.

I was thinking that maybe we can do something like this:
```
IIssueViewModel
{
 	IAnalysisIssueVisualization TaintIssueViz { get; }
        int Line { get; }
        int Column { get; }
        string DisplayPath { get; }
}

IIssuesControlViewModel<T> where T : IIssueViewModel
{
	ICommand NavigateCommand { get; }
        ObservableCollection<T> Issues { get; }
}

IHotspotViewModel : IIssueViewModel
{
	string CategoryDisplayName { get; }
}

IHotspotsControlViewModel : IIssuesControlViewModel<IHotspotViewModel>
{
	IHotspotViewModel SelectedHotspot { get; }
	ICommand RemoveCommand { get; }
}
```

And create a base class that can contain essentially everything that is in TaintIssuesControlViewModel.
WDYT?

File: src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.cs (hunk 1 above)
Comment (rita-g-sonarsource) [reply src/IssueViz.Security.UnitTests/Taint/TaintList/TaintIssuesControlViewModelTests.cs:35]: See my comment regarding the generic component. I think it's too much duplication.

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs (hunk 6 above)
Comment (duncanp-sonar) [reply src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:49]: Yes, I think there is too much duplication and there should be a couple of view model base classes (that's what I was alluding to when I mentioned identifying opportunities for refactoring above - perhaps I should have been more explicit). My plan was to make the changes to the taint tool window and control to get a working end-to-end where both view models are actually being consumed and then refactor to extract the common behaviour.

**Review by rita-g-sonarsource — APPROVED** [review]

File: src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs (hunk 6 above)
Comment (duncanp-sonar) [reply src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs:49]: FYI one part of the potentially common view model behaviour that I'm not currently sure about is selection, more specifically integration with the selection service. We currently have two selection services, `IAnalysisIssueSelectionService` and `IHotspotSelectionService`. I don't think we want to add a third one for taint. We might end up with one service that can be consumed in a consistent way by the editor light bulbs, flow window and error/hotspot/taint lists.

**Pull request merged**